### DISCUSSION TITLE
Does `generateStaticParams api` work only in server component?

### DISCUSSION THREAD

**Discussion by hyeongrok7874** [vercel/next.js#50603] [App Router] [answered]
I am currently unable to use the server component because I am using emotion.
Currently, "use client" is declared in page.js.

I want to use `generateStaticParams api` to generate pages through Dynamic Routes and generate pages at build time.

However, I don't know how to check if `generateStaticParams api` works well, and 404 is not returned even if `dynamicParams` route segment option is set to false.

So, I wonder if the `generateStaticParams api` or route segment option only works in the server component.

**Comment by devjiwonchoi** [comment, marked as answer]
Yes, `generateStaticParams` only works on Server Components since it is a `Server Function`.
It is directly stated [here](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes#generating-static-params).

> See [generateStaticParams server function documentation](https://nextjs.org/docs/app/api-reference/functions/generate-static-params) for more information and advanced use cases.

In production, `generateStaticParams` run at build time for [Dynamic Route](https://nextjs.org/docs/app/building-your-application/routing/dynamic-routes) which occurs on the server side.

>During next dev, generateStaticParams will be called when you navigate to a route.
During next build, generateStaticParams runs before the corresponding Layouts or Pages are generated.
During revalidation (ISR), generateStaticParams will not be called again.

In summary, you cannot use `generateStaticParams` in Client Components.

I would recommend you set `emotion` inside the `layout.js` as a Client Component, 
and leave the Dynamic Segment as Server Component in order to use the `generateStaticParams`.

Also, note that using `emotion` is not recommended yet to use in `App Router` since it is not fully compatible.

> The following are currently working on support:
[emotion](https://github.com/emotion-js/emotion/issues/2928)
[Material UI](https://github.com/mui/material-ui/issues/34905#issuecomment-1401306594)
[Chakra UI](https://github.com/chakra-ui/chakra-ui/issues/7180)

I would recommend you use one of these for the Client Component for now.

> The following libraries are supported in Client Components in the app directory:
[styled-jsx](https://nextjs.org/docs/app/building-your-application/styling/css-in-js#styled-jsx)
[styled-components](https://nextjs.org/docs/app/building-your-application/styling/css-in-js#styled-components)
[tamagui](https://tamagui.dev/docs/guides/next-js#server-components)
[style9](https://github.com/johanholmerin/style9)
[vanilla-extract](https://github.com/SuttonJack/vanilla-extract-app-dir)

**Reply by hyeongrok7874** [reply]
Your answer has been of great help to me.

Thank you very much again.
